File: leetcode/00003. Longest Substring Without Repeating Characters/C++/sol.cpp
// Sliding Window Approach with HashSet
class Solution {
    public:
    int lengthOfLongestSubstring(string s) {
        int ls = 0, left = 0, right = 0;
        unordered_set<char> us;
        while (left < s.length()) {
            while(right < s.length() && us.find(s[right]) == us.end()) {
                us.insert(s[right]);
                right++;             
            }
            ls = max(ls, right - left);
            us.erase(s[left]);
            left++;
        }
        return ls;
    }
};

// Sliding Window Approach with HashMap - Faster
class Solution {
    public:
    int lengthOfLongestSubstring(string s) {
        int ls = 0, left = 0, right = 0;
        unordered_map<char, int> um;
        while (right < s.length()) {
            if (um.find(s[right]) != um.end()) {
                left = max(left, um[s[right]] + 1);
            }
            um[s[right]] = right;
            ls = max(ls, right - left + 1);
            right++;
        }
        return ls;
    }
};

File: leetcode/00003. Longest Substring Without Repeating Characters/Python/sol.py
# Sliding Window Approach using Set
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        st = set()
        l = r = ls = 0
        while(l < len(s)):
            while(r < len(s) and s[r] not in st):
                st.add(s[r])
                r+=1
            ls = max(ls, r-l)
            st.remove(s[l])
            l+=1
        return ls
    
# Sliding Window Approach using Dictionary - Faster
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s:
            return 0
        max_length = 0
        left = 0
        char_index = {}
        for right, char in enumerate(s):
            if char in char_index and char_index[char] >= left:
                left = char_index[char] + 1
            if max_length < right - left + 1:
                max_length = right - left + 1
            char_index[char] = right
        return max_length